### PR TITLE
Add multiple layers selection

### PR DESCRIPTION
**Pull request by Jatana**

### CODE REVIEW

**Comment by Jatana** [comment]
ezgif-1-eb7ad4172f06

**Comment by thatbrguy** [comment]
Hey @Jatana, good work. Couple of things:

- Resolve the webpack errors that i've marked in the review.
- I'm noticing a weird behavior. When I click on a layer, it is redirecting me to `fabrik.cloudcv.org`, without open the layer details panel. I was not logged in when this happened. Can you verify if this happens to you too? I cannot test the multiple layer selection functionality as this happens.

File: ide/static/js/canvas.js [hunk 1]
```diff
@@ -235,7 +236,14 @@ class Canvas extends React.Component {
   allowDrop(event) {
     event.preventDefault();
   }
+  clickMultipleLayerEvent(event, layerId) {
+    event.preventDefault();
+    if (event.target.tagName.toLowerCase() == 'div') {
+      this.props.addLayerToMultipleSelection(layerId);
```
Comment (thatbrguy) [inline ide/static/js/canvas.js]: Webpack error:
`242:18  error  'addLayerToMultipleSelection' is missing in props validation  react/prop-types`

File: ide/static/js/content.js [hunk 2]
```diff
@@ -383,32 +385,60 @@ class Content extends React.Component {
...
+    if (this.state.selectedLayer == layerId) {
+      cl += ' selected';
+    }
+    if (this.state.hoveredLayer == layerId) {
+      cl += ' hover';
+    }
+    if (this.state.selectedLayers.indexOf(layerId) != -1) {
+      cl += ' multi-selected';
+    }
+    this.state.net[layerId].info.class = cl;
```
Comment (thatbrguy) [inline ide/static/js/content.js]: Webpack error (multiple lines):
```
  399:5  error  Do not mutate state directly. Use setState()  react/no-direct-mutation-state
  423:7  error  Do not mutate state directly. Use setState()  react/no-direct-mutation-state
  433:5  error  Do not mutate state directly. Use setState()  react/no-direct-mutation-state
  439:5  error  Do not mutate state directly. Use setState()  react/no-direct-mutation-state
```

File: ide/static/js/layer.js [hunk 3]
```diff
@@ -90,6 +90,7 @@ class Layer extends React.Component {
         }}
         data-type={this.props.type}
         onClick={(event) => this.props.click(event, this.props.id)}
+        onContextMenu={(event) => this.props.onContextMenu(event, this.props.id)}
```
Comment (thatbrguy) [inline ide/static/js/layer.js]: Webpack error:
`93:46  error  'onContextMenu' is missing in props validation  react/prop-types`

**Comment by Ram81** [comment]
@Jatana I see same behaviour as stated by @thatbrguy can you mention how does your current implementation allows us to choose multilpe layers.

**Comment by Ram81** [comment]
@Jatana can you resolve the conflicts before In review the PR.

File: ide/static/css/style.css [hunk 4]
```diff
@@ -174,6 +178,19 @@ input#netName.hidden {
...
 
+#layers-delete {
+  margin: 0px;
+  position: fixed;
+  top: 4%;
+  left: 22%;
+  width: 180px;
+}
+
+#layers-delete  > p{
```
Comment (Ram81) [inline ide/static/css/style.css]: Add a space after `p`

**Comment by Ram81** [comment]
@Jatana few comments:

- Delete button is at wrong place, please place it somewhere in sidebar after multiple layers are selected.

- After deleting random nodes from graph, I can't delete the dangling layers (i.e. one with no input and no output).

- Can you figure out some better way to select multiple layers `Ctrl + Click` is not that intuitive.

- I've noticed some unusual behaviour, first I opened a sidebar of a layer. Then without closing the side pane I started selecting multiple layers for deletion and then deleted layers. After deletion I wan't able to click anywhere and not close any UI component.

File: ide/static/js/canvas.js [hunk 5]
```diff
@@ -237,7 +238,14 @@ class Canvas extends React.Component {
   allowDrop(event) {
     event.preventDefault();
   }
+  clickMultipleLayerEvent(event, layerId) {
+    event.preventDefault();
+    if (event.target.tagName.toLowerCase() == 'div') {
```
Comment (Ram81) [inline ide/static/js/canvas.js]: Isn't there a better way to detect whether the event of click occurred on a layer component?

File: ide/static/js/canvas.js [hunk 6]
```diff
@@ -482,6 +493,14 @@ class Canvas extends React.Component {
       >
         {layers}
       </div>
+      <div id='layers-delete' className={`${this.props.getDeleteButtonStyle()}`} onClick={ () => this.props.smartDeleteLayers() }>
```
Comment (Ram81) [inline ide/static/js/canvas.js]: We need to place it some appropriate position and not on canvas.

**Comment by yashdusing** [comment]
Hey @Jatana ! What's the button click for selecting the layers ? I've tried shift+click, ctrl+click and both haven't worked for me

File: ide/static/js/topBar.js [hunk 7]
```diff
@@ -106,6 +106,18 @@ class TopBar extends React.Component {
             </div>
             {content}
         </div>
+        <div className="topbar-row">
```
Comment (Ram81) [inline ide/static/js/topBar.js]: Can you fix the alignment of the new icon, it doesn't look good at the center.

**Comment by Ram81** [comment]
@Jatana there is one more issue related to deselecting logic. Let's say I first select two layers for deletion and then I click on some other layer (without using shift) then deselection doesn't happen and modal overlaps with sidebar.
screen shot 2018-11-13 at 9 40 05 pm

**Comment by Ram81** [comment]
@Jatana there are still implementation issues in the PR please fix those and try to do a complete test of all the features to see if anything else is breaking or not because of your change. Let us know if you face any issues.

**Comment by Jatana** [comment]
I fix deselection logic, the problem with appearing errors when deleting layers is happening with standard delete method too. So, I think it is not my fault.

**Comment by Ram81** [comment]
@Jatana I still can't add new layers after deleting multiple layers. Please try to fix the issue.

**Comment by Ram81** [comment]
Closing this as student has abandoned the task.

**Pull request closed without merging**